### PR TITLE
build-tools: Add `bump deps` CLI command

### PR DESCRIPTION
**Pull request by tylerbutler**
_Update 2022-07-22: #11215 is merged so this is ready for review again._

_Update 2022-07-21: I'm going to wait until #11215 is merged, then rebase on top of it._

---

The `bump deps` command is used to bump the dependency ranges of release groups or individual packages. It's easiest to
understand with an example.

Consider this section of package.json.

```json
"dependencies": {
    "@releasegroup1/app": "^1.3.0",
    "@releasegroup1/lib": "^1.3.0",
    "@standalone/common-tools": "^1.24.0",
    "@standalone/eslint-config": "~1.28.2"
}
```

All of the dependencies are in the same repo. The first two dependencies listed are in a single release group, while the
other two are standalone packages.

When releasegroup1 publishes a prerelease version 1.4.0-xxxxxx, we would want to bump the dependency range in the
package above to be `^1.4.0-0`, which will pick up the new release. Doing that in one package with a release group that
has only two packages is straightforward. However, when a repo has dozens or hundreds of packages with lots of large
release groups, doing it manually becomes untenable.

The `bump deps` command automates this process. In the case above, we could use the following command to bump
releasegroup1 dependencies to `^1.4.0-0`:

```
flub bump deps releasegroup1 --version '^1.4.0-0'
```

That command will update the package.json like so:

```json
"dependencies": {
    "@releasegroup1/app": "^1.4.0-0",
    "@releasegroup1/lib": "^1.4.0-0",
    "@standalone/common-tools": "^1.24.0",
    "@standalone/eslint-config": "~1.28.2"
}
```

## Bumping based on current dependency range

It is very helpful to bump a dependency based on its current value and a bump type, such as "major" or "minor". The
following command yields the same results as the above command:

```shell
flub bump deps releasegroup1 --bumpType minor --prerelease
```

To bump to a release version instead, omit the `--prerelease` argument.

## Bumping standalone dependencies

Some packages are versioned independently from other release groups. In the example above, we could bump to the next
major version of the eslint-config package across the whole repo using the following command:

```shell
flub bump deps @standalone/eslint-config --bumpType major
```

That command will update the package.json like so:

```json
"dependencies": {
    "@releasegroup1/app": "^1.3.0",
    "@releasegroup1/lib": "^1.3.0",
    "@standalone/common-tools": "^1.24.0",
    "@standalone/eslint-config": "~2.0.0"
}
```

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -11,12 +11,25 @@
             "skipFiles": [
                 "<node_internals>/**"
             ],
+            "outFiles": [
```
Comment (tylerbutler) [inline .vscode/launch.json:14]: Revert these changes.

File: build-tools/packages/build-cli/README.md [hunk 2]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+    "@releasegroup1/lib": "^1.3.0",
+    "@standalone/common-tools": "^1.24.0",
+    "@standalone/eslint-config": "~1.28.2"
+}
+```
+
+All of the dependencies are in the same repo. The first two dependencies listed are in a single release group, while the
+other two are standalone packages.
+
+When releasegroup1 publishes a prerelease version 1.4.0-xxxxxx, we would want to bump the dependency range in the
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:37]:
```suggestion
When releasegroup1 publishes a prerelease version 1.4.0-xxxxxx, we want to bump the dependency range in the
```

File: build-tools/packages/build-cli/README.md [hunk 3]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+understand with an example.
+
+Consider this section of package.json.
+
+```json
+"dependencies": {
+    "@releasegroup1/app": "^1.3.0",
+    "@releasegroup1/lib": "^1.3.0",
+    "@standalone/common-tools": "^1.24.0",
+    "@standalone/eslint-config": "~1.28.2"
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:30]:
```suggestion
    "@standalone/common-tools": "~1.24.0",
    "@standalone/eslint-config": "^1.28.2"
```

File: build-tools/packages/build-cli/README.md [hunk 4]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+}
+```
+
+All of the dependencies are in the same repo. The first two dependencies listed are in a single release group, while the
+other two are standalone packages.
+
+When releasegroup1 publishes a prerelease version 1.4.0-xxxxxx, we want to bump the dependency range in the
+package above to be `^1.4.0-0`, which will pick up the new release. Doing that in one package with a release group that
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:38]:
```suggestion
package above to be `~1.4.0-0`, which will pick up the new release. Doing that in one package with a release group that
```

File: build-tools/packages/build-cli/README.md [hunk 5]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+
+The `bup deps` command is used to bump the dependency ranges of release groups or individual packages. It's easiest to
+understand with an example.
+
+Consider this section of package.json.
+
+```json
+"dependencies": {
+    "@releasegroup1/app": "^1.3.0",
+    "@releasegroup1/lib": "^1.3.0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:28]:
```suggestion
    "@releasegroup1/app": "~1.3.0",
    "@releasegroup1/lib": "~1.3.0",
```

File: build-tools/packages/build-cli/README.md [hunk 6]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+release groups, doing it manually becomes untenable.
+
+The `bump deps` command automates this process. In the case above, we could use the following command to bump
+releasegroup1 dependencies to `^1.4.0-0`:
+
+```
+flub bump deps releasegroup1 --version '^1.4.0-0'
+```
+
+That command will update the package.json like so:
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:49]:
```suggestion
The `bump deps` command automates this process. In the case above, we could use the following command to bump
releasegroup1 dependencies to `~1.4.0-0`:

```
flub bump deps releasegroup1 --version '~1.4.0-0'
```

That command will update the package.json like so:
```

File: build-tools/packages/build-cli/README.md [hunk 7]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+
+```json
+"dependencies": {
+    "@releasegroup1/app": "^1.4.0-0",
+    "@releasegroup1/lib": "^1.4.0-0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:54]:
```suggestion
    "@releasegroup1/app": "~1.4.0-0",
    "@releasegroup1/lib": "~1.4.0-0",
```

File: build-tools/packages/build-cli/README.md [hunk 8]
```diff
@@ -8,80 +8,155 @@ flub is not built in CI. You need to build it locally.
...
+```
+flub bump deps @standalone/eslint-config --bumpType major
+```
+
+That command will update the package.json like so:
+
+```json
+"dependencies": {
+    "@releasegroup1/app": "^1.3.0",
+    "@releasegroup1/lib": "^1.3.0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:85]:
```suggestion
    "@releasegroup1/app": "~1.3.0",
    "@releasegroup1/lib": "~1.3.0",
```

File: build-tools/packages/build-cli/src/base.ts [hunk 9]
```diff
@@ -49,4 +50,39 @@ export abstract class BaseCommand extends Command {
...
+    logWarn(message?: string, preMessage = "WARNING", ...args: unknown[]): void {
+        this.log(chalk.yellow(`${preMessage}: ${message}`), ...args);
+    }
+}
+
+/**
+ * A base class that should be used by commands that possibly modify repo state. Such commands should provide a
+ * consistent set of flags to control which checks should be run or skipped.
+ */
+export abstract class RepoStateModifyingCommand extends BaseCommand {
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/base.ts:63]: Unused currently, but I expect to use it in a future PR.

File: build-tools/packages/build-cli/src/commands/bump/deps.ts [hunk 10]
```diff
@@ -3,25 +3,293 @@
...
+        {
+            name: "package_or_release_group",
+            required: true,
+            description:
+                "The name of a package or a release group. Dependencies on these packages will be bumped.",
+        },
+    ];
 
     static flags = {
+        version: semverRangeFlag({
```
Comment (sonalivdeshpande) [inline build-tools/packages/build-cli/src/commands/bump/deps.ts:43]: do you plan to have all the flags under one file and reuse it as needed?

File: build-tools/packages/build-cli/README.md [hunk 11]
```diff
@@ -8,80 +8,150 @@ flub is not built in CI. You need to build it locally.
...
 * [Commands](#commands)
+* [Usage](#usage)
+* [Command reference](#command-reference)
 <!-- tocstop -->
 
+# Commands
+
+## bump deps
+
+The `bup deps` command is used to bump the dependency ranges of release groups or individual packages. It's easiest to
```
Comment (Josmithr) [inline build-tools/packages/build-cli/README.md:20]:
```suggestion
The `bump deps` command is used to bump the dependency ranges of release groups or individual packages. It's easiest to
```

File: build-tools/README.md [hunk 12]
```diff
@@ -16,6 +17,11 @@ policy-check, fluid-bump-version, etc.
 Note: Don't add new CLI commands to this package. Instead, add a new command to the `build-cli` package and import the
 functionality you need from this package.
 
+## @fluid-tools/version-tools
+
+This package contains APIs and a CLI for working with semantic versioning version strings and ranges, especially those
+using [Fluid-specific version schemes.](./packages/version-tools/README.md#version-schemes)
```
Comment (Josmithr) [inline build-tools/README.md:23]: (Not an issue for this PR) Do we currently do link validation on links in READMEs like this? Or do we just validate the website currently?

File: build-tools/packages/build-cli/README.md [hunk 13]
```diff
@@ -8,80 +8,150 @@ flub is not built in CI. You need to build it locally.
...
+```
+
+To bump to a release version instead, omit the `--prerelease` argument.
+
+## Bumping standalone dependencies
+
+Some packages are versioned independently from other release groups. In the example above, we could bump to the next
+major version of the eslint-config package across the whole repo using the following command:
+
+```
```
Comment (Josmithr) [inline build-tools/packages/build-cli/README.md:71]:
```suggestion
```shell
```

File: build-tools/packages/build-cli/src/base.ts [hunk 14]
```diff
@@ -49,4 +50,39 @@ export abstract class BaseCommand extends Command {
 
         return this._context;
     }
+
+    logWarn(message?: string, preMessage = "WARNING", ...args: unknown[]): void {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/base.ts:54]: (Not for this PR) I wonder if we should add a console utilities package or something. I could see utilities like this being useful for docs and test utils and such.

File: build-tools/packages/build-cli/src/commands/bump/deps.ts [hunk 15]
```diff
@@ -3,25 +3,293 @@
...
+            description:
+                "Bump dependencies on server packages to the current version, replacing any pre-release ranges with release ranges.",
+            command: "<%= config.bin %> <%= command.id %> server -g client -t current",
+        },
+    ];
+
+    private readonly finalMessages: string[] = [];
+
+    /**
+     *
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/bump/deps.ts:110]: TODO 😀

File: build-tools/packages/build-cli/src/commands/bump/deps.ts [hunk 16]
```diff
@@ -3,25 +3,293 @@
...
+
+        if (this.finalMessages.length > 0) {
+            this.log("=".repeat(72));
+            for (const msg of this.finalMessages) {
+                this.log(msg);
+            }
+        }
+    }
+}
+interface PackageWithRangeSpec {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/bump/deps.ts:231]: Nit: newline above and docs?

File: build-tools/packages/build-tools/src/bumpVersion/bumpDependencies.ts [hunk 17]
```diff
@@ -18,69 +19,119 @@ import { FluidRepo } from "../common/fluidRepo";
...
  * @param context - The repo {@link Context}.
- * @param commitMessage - The commit message to use when committing changes. This must be provided if `commit` is true.
- * @param commit - If true, commit the changes to a new branch.
  * @param bumpDepPackages - A map of package names to the version that dependencies on that package should be set to.
- * @param updateLock - If true, update the lock file by running `npm install` automatically.
- * @param release - If true, make dependencies target release version instead of pre-release versions (e.g. ^0.16.0 vs
- * ^0.16.0-0)
- * @param releaseGroup - Only update dependencies within this release group. If not specified, operates on the whole
- * repo.
+ * @param releaseGroup - Only update dependencies within this release group. To operates on the whole repo, pass
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/bumpVersion/bumpDependencies.ts:23]:
```suggestion
 * @param releaseGroup - Only update dependencies within this release group. To operate on the whole repo, pass
```

File: build-tools/packages/build-tools/src/bumpVersion/bumpDependencies.ts [hunk 18]
```diff
@@ -18,69 +19,119 @@ import { FluidRepo } from "../common/fluidRepo";
...
+            break;
+        }
+        // case "internal":
+        // case "virtualPatch":
+        default: {
+            fatal("NOT YET IMPLEMENTED");
+        }
+    }
+
+    // For each package in the repo, check its dependencies to see if they're in the list of packages to bump.
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/bumpVersion/bumpDependencies.ts:60]: TSDoc comment?

File: build-tools/packages/build-tools/src/bumpVersion/bumpVersionCli.ts [hunk 19]
```diff
@@ -298,7 +298,8 @@ async function main() {
                 await bumpDependencies(
                     context,
                     paramBumpDepPackages,
-                    /*updateLock*/ paramLocal,
+                    undefined,
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/bumpVersion/bumpVersionCli.ts:301]: Nit: I like the pattern you're using of adding param name comments before for ones that are potentially ambiguous to the reader. Would it be worth doing this for the undefined ones too?

File: tools/pipelines/build-build-tools.yml [hunk 20]
```diff
@@ -84,12 +84,12 @@ extends:
         customRegistry: 'useNpmrc'
         workingDir: '$(Build.SourcesDirectory)/build-tools'
     # Build the build-tools locally so it can be used to generate the monorepo package.json
-    - task: Npm@1
-      displayName: 'npm run build:compile'
-      inputs:
-        command: 'custom'
-        customCommand: 'run build:compile'
-        workingDir: '$(Build.SourcesDirectory)/build-tools'
+    # - task: Npm@1
```
Comment (Josmithr) [inline tools/pipelines/build-build-tools.yml:87]: Did you mean to leave this commented out?

**Comment by sonalivdeshpande** [comment]
would the same command get tweaked later to bump ranges `(>=2.0.0-internal.1.0.0 <2.0.0-internal.2.0.0)` as well for major/minor/patch releases?

File: build-tools/README.md (hunk 12 above)
Comment (tylerbutler) [reply build-tools/README.md:23]: Only the website, but there is a markdownlint config file in the root where I started to set some linting up a few years ago. Also I don't think the linting in the docs folder itself is run in CI. :( #9978

File: build-tools/packages/build-cli/src/base.ts (hunk 14 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/base.ts:54]: Yup, I've been thinking the same thing. I'll add it once there's enough stuff to give it a clear "identity" (I don't want it to be a "whatever we needed to share" lib :smile:)

File: build-tools/README.md (hunk 12 above)
Comment (Josmithr) [reply build-tools/README.md:23]: Good to know. I'm collecting misc. items for improving docs directory reliability for the larger website refactoring effort. I'll reference that there.

Comment (Josmithr) [reply build-tools/README.md:23]: Added an ADO item that points to the above issue.
https://dev.azure.com/fluidframework/internal/_workitems/edit/1366

File: build-tools/packages/build-cli/src/commands/bump/deps.ts (hunk 10 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/bump/deps.ts:43]: Yeah, the shared flag types are in the flags.ts file. But individual commands might need to override the default values, like I do here. For example, the shortcut char might conflict with an existing flag.

**Comment by tylerbutler** [comment]
> would the same command get tweaked later to bump ranges `(>=2.0.0-internal.1.0.0 <2.0.0-internal.2.0.0)` as well for major/minor/patch releases?

Yeah, but fortunately that will be pretty straightforward since the handling of the version schemes is all centralized now.

File: build-tools/packages/build-cli/src/commands/bump/deps.ts [hunk 21]
```diff
@@ -3,25 +3,300 @@
...
+    isVersionBumpTypeExtended,
+    FluidRepo,
+} from "@fluidframework/build-tools";
+import { incRange } from "@fluid-internal/version-tools";
+import { bumpTypeFlag, releaseGroupFlag, semverRangeFlag } from "../../flags";
+import { BaseCommand } from "../../base";
 
 /**
- * The `bump deps` command. This command is equivalent to `fluid-bump-version --dep`.
+ * Update the dependency version of a specified package or release group. That is, if package B depends on package A,
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/bump/deps.ts:21]: The second sentence here was a bit confusing to me on first read. I think maybe something like:
```md
That is, if one or more packages in the specified release group depend on package A, then this command...
```
would be more clear?

**Review by Josmithr — APPROVED** [review]

File: build-tools/packages/build-cli/src/commands/bump/deps.ts (hunk 21 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/bump/deps.ts]: Thanks; I think my latest update is much better.
